File: backend/src/schemas/handoff.py
from typing import List, Optional
from pydantic import BaseModel, Field


HANDOFF_STATUS_PATTERN = "^(requested|generating_summary|ready|accepted|completed|failed)$"
HANDOFF_REASON_PATTERN = "^(quota_exceeded|error|quality_issue|role_mismatch|manual|context_limit|other)$"
HANDOFF_RESULT_PATTERN = "^(success|partial|failed)$"
TASK_STATUS_PATTERN = "^(running|failed|handoff|completed|blocked)$"


class HandoffSummary(BaseModel):
    original_goal: str
    current_task: str
    completed_work: List[str]
    unfinished_work: List[str]
    important_constraints: List[str]
    key_decisions: List[str]
    errors_and_risks: List[str]
    next_suggested_steps: List[str]
    context_needed: List[str]


class HandoffTriggerRequest(BaseModel):
    to_agent_id: str = Field(..., min_length=1)
    to_model_id: Optional[str] = None
    reason: str = Field(..., pattern=HANDOFF_REASON_PATTERN)
    reason_description: Optional[str] = None
    include_recent_logs: bool = True
    include_current_output: bool = True


class HandoffAcceptRequest(BaseModel):
    agent_id: Optional[str] = None
    model_id: Optional[str] = None
    auto_continue: bool = True


class HandoffResultRequest(BaseModel):
    result_after_handoff: str = Field(..., pattern=HANDOFF_RESULT_PATTERN)
    status: Optional[str] = Field(default="completed", pattern="^(completed)$")
    result_note: Optional[str] = None


class HandoffTaskCreate(BaseModel):
    goal_id: str = Field(..., min_length=1)
    title: str = Field(..., min_length=1, max_length=255)
    description: Optional[str] = ""
    status: str = Field(default="running", pattern=TASK_STATUS_PATTERN)
    assigned_agent_id: str = Field(..., min_length=1)
    assigned_model_id: str = Field(..., min_length=1)
    current_output: Optional[str] = None
    error_message: Optional[str] = None


class HandoffTaskResponse(BaseModel):
    id: str
    goal_id: str
    title: str
    description: str
    status: str
    assigned_agent_id: str
    assigned_model_id: str
    assigned_worker_id: Optional[str]
    current_output: Optional[str]
    error_message: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]


class WorkerSessionResponse(BaseModel):
    id: str
    agent_id: str
    model_id: str
    goal_id: str
    task_id: str
    inherited_from_handoff_id: Optional[str]
    status: str
    current_context: Optional[str]
    final_output: Optional[str]
    error_message: Optional[str]
    created_at: Optional[str]
    updated_at: Optional[str]


class ExecutionLogResponse(BaseModel):
    id: str
    goal_id: Optional[str]
    task_id: Optional[str]
    agent_id: Optional[str]
    worker_id: Optional[str]
    model_id: Optional[str]
    handoff_id: Optional[str]
    level: str
    action: str
    message: str
    created_at: Optional[str]


class HandoffRecordResponse(BaseModel):
    id: str
    goal_id: str
    task_id: str
    from_agent_id: str
    from_model_id: str
    from_worker_id: Optional[str]
    to_agent_id: str
    to_model_id: str
    to_worker_id: Optional[str]
    reason: str
    reason_description: Optional[str]
    handoff_summary: HandoffSummary
    status: str
    result_after_handoff: Optional[str]
    result_note: Optional[str]
    tokens_before_handoff: int
    tokens_after_handoff: int
    time_saved_estimate_ms: Optional[int]
    error_message: Optional[str]
    created_at: Optional[str]
    summary_generated_at: Optional[str]
    accepted_at: Optional[str]
    completed_at: Optional[str]
    updated_at: Optional[str]
    task: Optional[HandoffTaskResponse] = None
    from_agent_name: Optional[str] = None
    from_agent_role: Optional[str] = None
    to_agent_name: Optional[str] = None
    to_agent_role: Optional[str] = None
    worker: Optional[WorkerSessionResponse] = None


class HandoffListItem(BaseModel):
    id: str
    goal_id: str
    task_id: str
    task_title: Optional[str]
    from_agent_id: str
    from_agent_name: Optional[str]
    from_model_id: str
    to_agent_id: str
    to_agent_name: Optional[str]
    to_model_id: str
    reason: str
    status: str
    result_after_handoff: Optional[str]
    created_at: Optional[str]


class HandoffListResponse(BaseModel):
    items: List[HandoffListItem]
    total: int
    page: int
    page_size: int


class HandoffTriggerResponse(BaseModel):
    handoff_id: str
    task_id: str
    status: str
    message: str


class HandoffAcceptResponse(BaseModel):
    handoff_id: str
    worker_id: str
    task_id: str
    status: str


class HandoffResultResponse(BaseModel):
    handoff_id: str
    status: str
    result_after_handoff: str


class ApiResponse(BaseModel):
    success: bool
    data: Optional[dict] = None
    message: Optional[str] = None

File: frontend/src/types/handoff.ts
export type HandoffStatus = 'requested' | 'generating_summary' | 'ready' | 'accepted' | 'completed' | 'failed';
export type HandoffReason = 'quota_exceeded' | 'error' | 'quality_issue' | 'role_mismatch' | 'manual' | 'context_limit' | 'other';
export type HandoffResult = 'success' | 'partial' | 'failed';

export interface HandoffSummary {
  original_goal: string;
  current_task: string;
  completed_work: string[];
  unfinished_work: string[];
  important_constraints: string[];
  key_decisions: string[];
  errors_and_risks: string[];
  next_suggested_steps: string[];
  context_needed: string[];
}

export interface HandoffTask {
  id: string;
  goal_id: string;
  title: string;
  description: string;
  status: string;
  assigned_agent_id: string;
  assigned_model_id: string;
  assigned_worker_id?: string | null;
  current_output?: string | null;
  error_message?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface WorkerSession {
  id: string;
  agent_id: string;
  model_id: string;
  goal_id: string;
  task_id: string;
  inherited_from_handoff_id?: string | null;
  status: string;
  current_context?: string | null;
  final_output?: string | null;
  error_message?: string | null;
  created_at?: string | null;
  updated_at?: string | null;
}

export interface HandoffRecord {
  id: string;
  goal_id: string;
  task_id: string;
  from_agent_id: string;
  from_model_id: string;
  from_worker_id?: string | null;
  to_agent_id: string;
  to_model_id: string;
  to_worker_id?: string | null;
  reason: HandoffReason;
  reason_description?: string | null;
  handoff_summary: HandoffSummary;
  status: HandoffStatus;
  result_after_handoff?: HandoffResult | null;
  result_note?: string | null;
  tokens_before_handoff: number;
  tokens_after_handoff: number;
  time_saved_estimate_ms?: number | null;
  error_message?: string | null;
  created_at?: string | null;
  summary_generated_at?: string | null;
  accepted_at?: string | null;
  completed_at?: string | null;
  updated_at?: string | null;
  task?: HandoffTask | null;
  from_agent_name?: string | null;
  from_agent_role?: string | null;
  to_agent_name?: string | null;
  to_agent_role?: string | null;
  worker?: WorkerSession | null;
}

export interface HandoffListItem {
  id: string;
  goal_id: string;
  task_id: string;
  task_title?: string | null;
  from_agent_id: string;
  from_agent_name?: string | null;
  from_model_id: string;
  to_agent_id: string;
  to_agent_name?: string | null;
  to_model_id: string;
  reason: HandoffReason;
  status: HandoffStatus;
  result_after_handoff?: HandoffResult | null;
  created_at?: string | null;
}

export interface HandoffListResponse {
  items: HandoffListItem[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface HandoffFilters {
  goal_id?: string;
  task_id?: string;
  reason?: string;
  status?: string;
  from_agent_id?: string;
  to_agent_id?: string;
  search?: string;
  page?: number;
  page_size?: number;
}

export interface HandoffTriggerRequest {
  to_agent_id: string;
  to_model_id?: string;
  reason: HandoffReason;
  reason_description?: string;
  include_recent_logs?: boolean;
  include_current_output?: boolean;
}

export interface HandoffAcceptRequest {
  agent_id?: string;
  model_id?: string;
  auto_continue?: boolean;
}

export interface HandoffResultRequest {
  result_after_handoff: HandoffResult;
  status?: 'completed';
  result_note?: string;
}

export interface ApiResponse<T> {
  success: boolean;
  data?: T;
  error?: {
    code: string;
    message: string;
  };
}

export const HANDOFF_STATUS_LABELS: Record<string, string> = {
  requested: '等待处理',
  generating_summary: '生成摘要中',
  ready: '可接手',
  accepted: '已接手',
  completed: '已完成',
  failed: '失败',
};

export const HANDOFF_REASON_LABELS: Record<string, string> = {
  quota_exceeded: '额度不足',
  error: '模型错误',
  quality_issue: '质量问题',
  role_mismatch: '角色不匹配',
  manual: '手动交接',
  context_limit: '上下文限制',
  other: '其他',
};

export const HANDOFF_RESULT_LABELS: Record<string, string> = {
  success: '成功',
  partial: '部分完成',
  failed: '失败',
};

export const HANDOFF_STATUS_CLASSES: Record<string, string> = {
  requested: 'bg-amber-100 text-amber-800 border-amber-200',
  generating_summary: 'bg-lavender-100 text-lavender-800 border-lavender-200',
  ready: 'bg-lavender-100 text-lavender-800 border-lavender-200',
  accepted: 'bg-sky-100 text-sky-700 border-sky-200',
  completed: 'bg-green-100 text-green-800 border-green-200',
  failed: 'bg-red-100 text-red-800 border-red-200',
};

export const HANDOFF_RESULT_CLASSES: Record<string, string> = {
  success: 'bg-green-100 text-green-800 border-green-200',
  partial: 'bg-amber-100 text-amber-800 border-amber-200',
  failed: 'bg-red-100 text-red-800 border-red-200',
};

export const HANDOFF_STATUS_DOTS: Record<string, string> = {
  requested: 'bg-amber-500',
  generating_summary: 'bg-lavender-500 animate-pulse',
  ready: 'bg-lavender-500',
  accepted: 'bg-sky-500',
  completed: 'bg-green-500',
  failed: 'bg-red-500',
};
